File: app/src/main/java/com/faizal/bottomnavigation/chats/ObservableSnapshotArray.java
package com.faizal.bottomnavigation.chats;

import com.faizal.bottomnavigation.chats.kotlin.BaseCachingSnapshotParser;
import com.faizal.bottomnavigation.chats.kotlin.ChangeEventListener;
import com.faizal.bottomnavigation.chats.kotlin.SnapshotParser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestoreException;

import androidx.annotation.NonNull;

/**
 * Subclass of {@link BaseObservableSnapshotArray} for Firestore data.
 */
public abstract class ObservableSnapshotArray<T>
        extends BaseObservableSnapshotArray<DocumentSnapshot, FirebaseFirestoreException, ChangeEventListener, T> {
    /**
     * @see BaseObservableSnapshotArray#BaseObservableSnapshotArray(BaseCachingSnapshotParser)
     */
    public ObservableSnapshotArray(@NonNull SnapshotParser<T> parser) {
        super(new CachingSnapshotParser<>(parser));
    }
}

File: app/src/main/java/com/faizal/bottomnavigation/chats/kotlin/ObservableSnapshotArray.kt
package com.faizal.bottomnavigation.chats.kotlin

import com.faizal.bottomnavigation.chats.BaseObservableSnapshotArray
import com.faizal.bottomnavigation.chats.CachingSnapshotParser
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestoreException

/**
 * Subclass of [BaseObservableSnapshotArray] for Firestore data.
 */
abstract class ObservableSnapshotArray<T>
/**
 * @see BaseObservableSnapshotArray.BaseObservableSnapshotArray
 */(parser: SnapshotParser<T?>) : BaseObservableSnapshotArray<DocumentSnapshot?, FirebaseFirestoreException?, ChangeEventListener?, T>(CachingSnapshotParser<T>(parser))